katdal/test/test_datasources.py

```diff
@@ -0,0 +1,113 @@
+
+
+def ramp(shape, offset=0.0, slope=1.0, dtype=np.float_):
+    """Generate a multidimensional ramp of values of the given dtype."""
+    x = offset + slope * np.arange(np.prod(shape), dtype=np.float_)
+    return x.astype(dtype).reshape(shape)
+
+
+def to_dask_array(x):
+    """Turn ndarray `x` into dask array with the standard vis-like chunking."""
+    n_corrprods = x.shape[2] if x.ndim >= 3 else x.shape[1] / 8
+    chunk_size = 4 * n_corrprods * 8
+    chunks = generate_chunks(x.shape, x.dtype, chunk_size,
+                             dims_to_split=(0, 1), power_of_two=True)
+    return da.from_array(x, chunks)
+
+
+def put_fake_dataset(store, base_name, shape):
+    """Write a fake dataset into the chunk store."""
+    data = {'correlator_data': ramp(shape, dtype=np.float32) * (1 - 1j),
+            'flags': np.zeros(shape, dtype=np.uint8),
+            'weights': ramp(shape, slope=256. / np.prod(shape), dtype=np.uint8),
+            'weights_channel': ramp(shape[:-1], dtype=np.float32)}
+    ddata = {k: to_dask_array(array) for k, array in data.items()}
+    chunk_info = {k: {'chunks': darray.chunks, 'dtype': darray.dtype,
+                      'shape': darray.shape} for k, darray in ddata.items()}
+    push = [store.put_dask_array(store.join(base_name, k), darray)
+            for k, darray in ddata.items()]
+    da.compute(push)
+    return data, chunk_info
+
+
+class TestChunkStoreVisFlagsWeights(object):
+    """Test the :class:`ChunkStoreVisFlagsWeights` dataset store."""
+
+    @classmethod
+    def setup_class(cls):
+        cls.tempdir = tempfile.mkdtemp()
+
+    @classmethod
+    def teardown_class(cls):
+        shutil.rmtree(cls.tempdir)
+
+    def test_construction(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb1'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        vfw = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+        weights = data['weights'] * data['weights_channel'][..., np.newaxis]
+        # Check that data is as expected when accessed via VisFlagsWeights
+        assert_equal(vfw.shape, data['correlator_data'].shape)
+        assert_array_equal(vfw.vis.compute(), data['correlator_data'])
+        assert_array_equal(vfw.flags.compute(), data['flags'])
+        assert_array_equal(vfw.weights.compute(), weights)
+
+    def test_missing_chunks(self):
+        # Put fake dataset into chunk store
+        store = NpyFileChunkStore(self.tempdir)
+        base_name = 'cb2'
+        shape = (10, 64, 30)
+        data, chunk_info = put_fake_dataset(store, base_name, shape)
+        # Delete a random chunk in each array of the dataset
+        missing_chunks = {}
+        for array, info in chunk_info.items():
+            array_name = store.join(base_name, array)
+            slices = da.core.slices_from_chunks(info['chunks'])
+            index = np.random.randint(len(slices))
```

Comment:
Use an np.RandomState with a fixed seed so that the test behaves the same every time. Or alternatively, hand-code the missing pieces. You probably want to avoid two of the arrays having the same piece go missing, since that would weaken coverage of the test on data_lost.

Reply:
By the way, I think you can just to `random.choice(slices)` to directly pick a random slice rather than first computing a random slice index and then indexing.

Reply:
Done.

I knew there was a better function but could not remember the name - thanks!

Reply:
Unfortunately `random.choice` doesn't work as the list of tuples of slices are interpreted as an N-D array with N > 1. Back to `randint` (but kept the `RandomState` idea).

Reply:
Switched back to `random.choice`, not `np.random.choice` :-)